src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,523 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// An endpoint used to acquire the diagnostics stream for a runtime instance.
+    /// </summary>
+    internal interface IIpcEndpoint
+    {
+        /// <summary>
+        /// Wait for an available diagnostics connection to the runtime instance.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>
+        /// A task the completes when a diagnostics connection to the runtime instance becomes available.
+        /// </returns>
+        Task WaitForConnectionAsync(CancellationToken token);
+
+        /// <summary>
+        /// Connects to the underlying IPC Transport and opens a read/write-able Stream
+        /// </summary>
+        /// <returns>A Stream for writing and reading data to and from the target .NET process</returns>
+        Stream Connect();
+    }
+
+    internal abstract class BaseIpcEndpoint : IIpcEndpoint
```

Comment:
It has gotten complicated. And if I can simplify it, I definitely will.

Here's why it is so complicated:
Let's think of the endpoint implementations as things that are trying to facilitate production of streams and consumption of streams.

For ServerIpcEndpoint, the producer of the stream is the ReversedDiagnosticsServer (it calls ServerIpcEndpoint.SetStream, which is non-async) and the consumers are the WaitForConnectionAsync (async) and Connect (non-async) methods. The WaitForConnectionAsync method is also a producer because it really doesn't consume the stream; it consumes it to check the connectedness of the stream and then puts it back to make it available again.

This would normally be a simple producer consumer problem (even if there are multiple producers and multiple consumers). There should only be at most one stream that is provided to the endpoint at any given time because the runtime shouldn't reconnect until a stream was consumed and started communicating a diagnostic request. Thus, there's a scalar value that in the endpoint rather than have a list/queue of values like in the traditional producer consumer problem. What we have is a queue of consumers (the StreamTarget class and derived classes) that either accept the stream async (WaitForConnectionAsync method) or non-async (Connect method). If it were just the producer (SetStream method) and the single consumer (Connect method), this would be much simpler. However, the WaitForConnectionAsync consumes the stream and then provides it again to the queue of consumers. It holds the steam semaphore (which is used as an async "auto" reset event) so that it can provide the same stream object before anything else can acquire the lock on the stream field and provide a new stream.

Maybe this can be simplified by not having the WaitForConnectionAsync hold the "lock" (the semaphore) while it is checking the state of the stream. The runtime instance shouldn't be reconnecting during this operation so there shouldn't be any concern for another stream being provided at the same time as checking the state of the current stream. With that in mind, I might be able to get rid of the StreamTarget concept and use SemaphoreSlim (allows both async and non-async waiting notions) instead such that consumers will register a new semaphore instance with the queue and producers will release the next semaphore in the queue after setting the stream field. The possible problem I see with this is that if a consumer cancels, then it's semaphore is still in the queue; producers won't know if the semaphore represents a valid waiting operation (I might be able to dispose the semaphore upon cancellation and then have the producer catch ObjectDisposedException, but that might require some synchronization there too because there can be a window of time between when cancellation occurs and the semaphore is disposed).

Reply:
I've kept the StreamTarget concept in order to maintain the synchronization of possible cancellation by the consumers vs providing the stream from producers but simplified it and the BaseIpcEndpoint class.